quickwit/quickwit-storage/src/opendal_storage/google_cloud_storage.rs

```diff
@@ -0,0 +1,158 @@
+    pub fn new(storage_config: GoogleCloudStorageConfig) -> Self {
+        Self { storage_config }
+    }
+}
+
+#[async_trait]
+impl StorageFactory for GoogleCloudStorageFactory {
+    fn backend(&self) -> StorageBackend {
+        StorageBackend::Google
+    }
+
+    async fn resolve(&self, uri: &Uri) -> Result<Arc<dyn Storage>, StorageResolverError> {
+        let storage = from_uri(&self.storage_config, uri)?;
+        Ok(Arc::new(DebouncedStorage::new(storage)))
+    }
+}
+
+/// Creates an emulated storage for testing.
+#[cfg(feature = "integration-testsuite")]
+pub fn new_emulated_google_cloud_storage(
+    uri: &Uri,
+) -> Result<OpendalStorage, StorageResolverError> {
+    let (bucket, root) = parse_google_uri(&uri).expect("must be valid google uri");
+
+    let mut cfg = opendal::services::Gcs::default();
+    cfg.bucket(&bucket);
+    cfg.root(&root.to_string_lossy());
+    // The default port for the fake gcs server is 4443.
+    cfg.endpoint("http://127.0.0.1:4443");
+
+    #[derive(Debug)]
+    struct DummyTokenLoader;
+    #[async_trait]
+    impl reqsign::GoogleTokenLoad for DummyTokenLoader {
+        async fn load(&self, _: reqwest::Client) -> anyhow::Result<Option<reqsign::GoogleToken>> {
+            Ok(Some(reqsign::GoogleToken::new(
+                "dummy",
+                86400,
+                "https://www.googleapis.com/auth/devstorage.full_control",
+            )))
+        }
+    }
+    cfg.customed_token_loader(Box::new(DummyTokenLoader));
+
+    let store = OpendalStorage::new_google_cloud_storage(uri.clone(), cfg)?;
+    Ok(store)
+}
+
+fn from_uri(
+    google_cloud_storage_config: &GoogleCloudStorageConfig,
+    uri: &Uri,
+) -> Result<OpendalStorage, StorageResolverError> {
+    let credential_path = google_cloud_storage_config
+        .resolve_credential_path()
+        .ok_or_else(|| {
+            let message = format!(
+                "could not find Google credential path in environment variable `{}` or storage \
+                 config",
+                GoogleCloudStorageConfig::GOOGLE_CLOUD_STORAGE_CREDENTIAL_PATH_ENV_VAR
+            );
+            StorageResolverError::InvalidConfig(message)
+        })?;
+    let (bucket_name, prefix) = parse_google_uri(uri).ok_or_else(|| {
+        let message = format!("failed to extract bucket name from google URI: {uri}");
+        StorageResolverError::InvalidUri(message)
+    })?;
+
+    let mut cfg = opendal::services::Gcs::default();
+    cfg.credential_path(&credential_path);
+    cfg.bucket(&bucket_name);
+    cfg.root(&prefix.to_string_lossy());
+
+    let store = OpendalStorage::new_google_cloud_storage(uri.clone(), cfg)?;
+    Ok(store)
+}
+
+fn parse_google_uri(uri: &Uri) -> Option<(String, PathBuf)> {
+    // Ex: gs://bucket/prefix.
+    static URI_PTN: OnceCell<Regex> = OnceCell::new();
+
+    let captures = URI_PTN
+        .get_or_init(|| {
+            Regex::new(r"gs(\+[^:]+)?://(?P<bucket>[^/]+)(/(?P<prefix>.+))?")
```

Comment:
nitpick. 

`gs://bucket/` works thanks to partial matching.

Maybe we could tweak it into:
```suggestion
            Regex::new(r"gs(\+[^:]+)?://(?P<bucket>[^/]+)(/(?P<prefix>.+)?)?$")
```
or
```suggestion
            Regex::new(r"gs(\+[^:]+)?://(?P<bucket>[^/]+)(/(?P<prefix>.*))?$")
```
(note the $ at the end)